servicetalk-http-api/src/main/java/io/servicetalk/http/api/BlockingStreamingHttpMessageBodyUtils.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.api;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.concurrent.BlockingIterator;
+
+import java.util.NoSuchElementException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.Nullable;
+
+import static java.lang.System.nanoTime;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+final class BlockingStreamingHttpMessageBodyUtils {
+    private BlockingStreamingHttpMessageBodyUtils() {
+    }
+
+    static BlockingStreamingHttpMessageBody<Buffer> newMessageBody(BlockingIterator<Object> rawMsgBody) {
+        return new BlockingStreamingHttpMessageBody<Buffer>() {
+            private final BlockingStreamingHttpMessageBodyAdapter<Object> msgBody =
+                    new BlockingStreamingHttpMessageBodyAdapter<>(rawMsgBody);
+
+            @Override
+            public BlockingIterator<Buffer> payloadBody() {
+                return msgBody;
+            }
+
+            @Nullable
+            @Override
+            public HttpHeaders trailers() {
+                return msgBody.trailers();
+            }
+        };
+    }
+
+    static <T> BlockingStreamingHttpMessageBody<T> newMessageBody(
+            BlockingIterator<Object> rawMsgBody, HttpHeaders headers, HttpStreamingDeserializer<T> deserializer,
+            BufferAllocator allocator) {
+        final BlockingStreamingHttpMessageBodyAdapter<Object> msgBody =
+                new BlockingStreamingHttpMessageBodyAdapter<>(rawMsgBody);
+        BlockingIterator<T> deserialized = deserializer.deserialize(headers, () -> msgBody, allocator).iterator();
+        return new BlockingStreamingHttpMessageBody<T>() {
+            @Override
+            public BlockingIterator<T> payloadBody() {
+                return deserialized;
+            }
+
+            @Nullable
+            @Override
+            public HttpHeaders trailers() {
+                return msgBody.trailers();
+            }
+        };
+    }
+
+    private static final class BlockingStreamingHttpMessageBodyAdapter<I> implements BlockingIterator<Buffer> {
```

Comment:
Should access to this class' state be made thread-safe? `BlockingIterator` comes from the `io.servicetalk.concurrent` package. I'd assume it can be used in a concurrent context. The parent class has no mention of it requiring single-thread access.

Reply:
No this isn't necessary. `BlockingIterator` is in `concurrent` package because it is how we convert to/from `java.util.Iterator` related types. This type just adds timeouts at the call site and makes iterable `Closable` (which will cancel the underlying publisher if it isn't fully consumed).